core/commands/add.go

```diff
@@ -88,6 +88,31 @@ Buzhash or Rabin fingerprint chunker for content defined chunking by
 specifying buzhash or rabin-[min]-[avg]-[max] (where min/avg/max refer
 to the desired chunk sizes in bytes), e.g. 'rabin-262144-524288-1048576'.
 
+If you just specify 'rabin' without parameters the default sizes are
+'rabin-87381-262144-393216' and the parameters used for buzhash are
+always min: 131072 and max: 524288 with a mask of 65536. Buzhash
```

Comment:
Note that these default sizes are far from optimal for deduplication or speed according to my testing at;

https://github.com/dbaarda/rollsum-chunking/blob/master/RESULTS.rst

For any rollsum using the normal chunker (ie, no fancy dynamic match like FastCDC or truncation handler like Regression Chunking) the optimal sizes for both deduplication and speed (from cut-point-skipping) for a given average chunk size is with min == avg, max>=8*avg. The real average block size is actually min+avg, so I prefer to use the term "target" or "tgt" instead of "avg", so that avg=min+tgt, with optimal settings of min=tgt=avg/2, max>=4*avg.

The code currently also seems to support specifying just `rabin-[avg]` which is the same as `rabin-[avg/3]-[avg]-[avg*1.5]`. Also, `<avg>` will get rounded-down to the nearest power of 2 value. There is also constraints that `min<avg<max<=1048576`

For Buzhash the `buzMask = 1<<17 - 1`. Looking at go operator precedence it seems `-` binds with a higher precedence than `<<`, so this really is `1<<16 = 65536`, which is a mask with only one bit set, equivalent to avg=2 for an average block size of 128K+2. This seems like a terrible bug, and the intention was really to have `buzMask = (1<<17) - 1' which is a mask with 17 bits set for avg=128K. This bug might explain why buzhash chunking deduplication has been so crap in testing; it is pretty much identical to `size-131074`

So the current defaults for rabin give an average block size of 341K, and buzhash (due to the bug mentioned above) of 128K+2. 

So better defaults for an average block size of  256K would be rabin-131071-131072-1048576, and for buzhash would be min=131072,max=1048576, mask=(1<<17) -1.

I know changing these defaults now will break deduplication against content already added with the old defaults, but given buzhash has effectively been broken by a bug fixing it is probably fine. Fixing rabin is more contentious, but you could make `rabin-[avg]` equivalent to `rabin-[avg/2-1]-[avg/2]-[min(avg*4,1M)]` so that it gives better defaults when specifying a target average size.

Reply:
Yeah I was wondering what the - 1 on the bitshift is supposed to do as well...

Buzhash isn't *that* terrible right now... It actually performed as well as rabin overall in my tests: https://github.com/ipfs/go-ipfs-chunker/pull/31#issuecomment-880262126

But maybe @kubuxu can give this a look? :)

Reply:
Actually, it seems `<<` does have higher precedence than `-` in go, so there is not a bug in buzhash.

https://go.dev/ref/spec#Operators

I was misled by incorrect documentation here;

https://www.tutorialspoint.com/go/go_operators_precedence.htm